### PR TITLE
add json.decode processor

### PR DESCRIPTION
**Pull request by maha-hajja**
### Description

Related to  #1290 

### Quick checks:

- [x] I have followed the [Code Guidelines](https://github.com/ConduitIO/conduit/blob/main/docs/code_guidelines.md).
- [x] There is no other [pull request](https://github.com/ConduitIO/conduit/pulls) for the same update/change.
- [x] I have written unit tests.
- [x] I have made sure that the PR is of reasonable size and can be easily reviewed.

### CODE REVIEW

File: pkg/plugin/processor/builtin/jsonDecode.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+type jsonDecodeConfig struct {
+	// Field is the target field, as it would be addressed in a Go template (e.g. `.Payload.After`).
+	// Applicable values are `.Key`, `.Payload.Before` and `.Payload.After`, as they accept structured data format.
+	Field string `json:"field" validate:"required,inclusion=.Key|.Payload.Before|.Payload.After"`
+}
+
+func (p *jsonDecode) Specification() (sdk.Specification, error) {
+	return sdk.Specification{
+		Name:    "json.decode",
+		Summary: "Decodes a specific field from raw data (string) to JSON structured data.",
```
Comment (lovromazgon) [inline pkg/plugin/processor/builtin/jsonDecode.go:47]:
```suggestion
		Summary: "Decodes a specific field from JSON raw data (string) to structured data.",
```

File: pkg/plugin/processor/builtin/jsonDecode.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+func newJSONDecode() *jsonDecode {
+	return &jsonDecode{}
+}
+
+type jsonDecodeConfig struct {
+	// Field is the target field, as it would be addressed in a Go template (e.g. `.Payload.After`).
+	// Applicable values are `.Key`, `.Payload.Before` and `.Payload.After`, as they accept structured data format.
+	Field string `json:"field" validate:"required,inclusion=.Key|.Payload.Before|.Payload.After"`
```
Comment (lovromazgon) [inline pkg/plugin/processor/builtin/jsonDecode.go:41]: I wouldn't be that strict, I'd also support anything under `.Key.*`, `.Payload.Before.*` and `.Payload.After.*`. Imagine pulling records from Postgres and one column contains JSON data that we want to parse, so `.Payload.After["my_json_column"]` is a valid input and a valid use case.

File: pkg/plugin/processor/builtin/jsonDecode.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+	for _, record := range records {
+		rec := record
+		ref, err := p.referenceResolver.Resolve(&rec)
+		if err != nil {
+			return append(out, sdk.ErrorRecord{Error: err})
+		}
+
+		data := ref.Get()
+		switch d := data.(type) {
+		case opencdc.RawData:
```
Comment (lovromazgon) [inline pkg/plugin/processor/builtin/jsonDecode.go:87]: If we add support for fields nested in the key or payload, we'll have to check for `[]byte` and `string` types as well.

File: pkg/plugin/processor/builtin/jsonDecode.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+	}
+	return out
+}
+
+func (p *jsonDecode) Teardown(context.Context) error {
+	return nil
+}
+
+func (p *jsonDecode) setJSONData(bytes []byte, ref sdk.Reference) error {
+	var jsonData opencdc.StructuredData
```
Comment (lovromazgon) [inline pkg/plugin/processor/builtin/jsonDecode.go:121]: This assumes that the JSON data is always an object (map), what if I want to decode an array? 😢

Comment (lovromazgon) [reply pkg/plugin/processor/builtin/jsonDecode.go]: I added the code myself, to keep this moving.

**Review by hariso — APPROVED** [review]

File: pkg/plugin/processor/builtin/jsonDecode.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:generate paramgen -output=jsonDecode_paramgen.go jsonDecodeConfig
+
+package builtin
```
[the remark this answers is not in the document]
Comment (lovromazgon) [reply pkg/plugin/processor/builtin/jsonDecode.go:17]: I agree, though I'm not sure how much that would interfere with the ongoing work from Maha, and given that we have that refactoring PR I'd say we can do it there.

**Review by hariso — APPROVED** [review]
Great team work here!

**Pull request merged**